### PR TITLE
Clean up network interface and public ip on failure

### PR DESCRIPTION
**Pull request by liu-samuel**
When vm provision fails, there was a bug where the network interface and public ip address were still present of the ui console. This manually deletes them.

Azure has a mandatory 3 minute timeout after creation before i can delete the partial data so I'm sleeping for 180 seconds for now and I'm not sure how to get around it.

EDIT: Changed the sleep 180 to call requeue phase to add delete_instance back onto the queue

Related: 
https://github.com/ManageIQ/manageiq/pull/23175

### CODE REVIEW

File: app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb [hunk 1]
```diff
@@ -6,4 +6,22 @@ def customize_destination
...
+    delete_instance(phase_context[:clone_options])
+  end
+
+  def delete_instance(instance_attrs)
+    source.with_provider_connection do |azure|
+      nis = ::Azure::Armrest::Network::NetworkInterfaceService.new(azure)
+      ips = ::Azure::Armrest::Network::IpAddressService.new(azure)
+      ni = nis.get(instance_attrs[:name], resource_group.name)
+      ip = ips.get("#{instance_attrs[:name]}-publicIp", resource_group.name)
+      sleep(180)
```
Comment (Fryguy) [inline app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb:22]: Wonder if there's a way we can leverage the queue and say *run this 3 minutes from now"

File: app/models/manageiq/providers/azure/cloud_manager/provision/cloning.rb [hunk 2]
```diff
@@ -183,5 +183,9 @@ def start_clone(clone_options)
 
       File.join(azure.subscription_id, vm.resource_group.downcase, vm.type.downcase, vm.name)
     end
+  rescue StandardError => e
+    return if phase_context[:clone_options].nil?
+    signal :clone_failure_cleanup
+    raise
```
Comment (agrare) [inline app/models/manageiq/providers/azure/cloud_manager/provision/cloning.rb:189]: `signal` is essentially a method call so this raise won't necessarily be reachable

File: app/models/manageiq/providers/azure/cloud_manager/provision/cloning.rb [hunk 3]
```diff
@@ -183,5 +183,9 @@ def start_clone(clone_options)
 
       File.join(azure.subscription_id, vm.resource_group.downcase, vm.type.downcase, vm.name)
     end
+  rescue StandardError => e
+    return if phase_context[:clone_options].nil?
```
Comment (agrare) [inline app/models/manageiq/providers/azure/cloud_manager/provision/cloning.rb:187]: Could we move this check into `clone_failure_cleanup`

File: app/models/manageiq/providers/azure/cloud_manager/provision/cloning.rb [hunk 4]
```diff
@@ -183,5 +183,9 @@ def start_clone(clone_options)
 
       File.join(azure.subscription_id, vm.resource_group.downcase, vm.type.downcase, vm.name)
     end
+  rescue StandardError => e
```
Comment (agrare) [inline app/models/manageiq/providers/azure/cloud_manager/provision/cloning.rb:186]: Can we rescue a more specific exception class?  Was there an AzureArmrest exception that was raised in this case?

File: app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb [hunk 5]
```diff
@@ -6,4 +6,21 @@ def customize_destination
 
     signal :post_create_destination
   end
+
+  def clone_failure_cleanup
+    delete_instance(phase_context[:clone_options])
```
Comment (agrare) [inline app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb:11]:
```suggestion
    delete_instance(phase_context[:clone_options]) if phase_context[:clone_options]
```

Comment (agrare) [reply app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb:11]: I think we should signal_abort after this is successful, so maybe something like
```suggestion
    begin
      delete_instance(phase_context[:clone_options])
      signal_abort
    rescue
      requeue_phase(3.minutes)
    end
```

File: app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb [hunk 6]
```diff
@@ -6,4 +6,21 @@ def customize_destination
...
+  def delete_instance(instance_attrs)
+    source.with_provider_connection do |azure|
+      nis = ::Azure::Armrest::Network::NetworkInterfaceService.new(azure)
+      ips = ::Azure::Armrest::Network::IpAddressService.new(azure)
+      ni = nis.get(instance_attrs[:name], resource_group.name)
+      ip = ips.get("#{instance_attrs[:name]}-publicIp", resource_group.name)
+      nis.delete(instance_attrs[:name], resource_group.name) if ni
+      ips.delete("#{instance_attrs[:name]}-publicIp", resource_group.name) if ip
+    end
+  rescue
```
Comment (agrare) [inline app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb:23]: Was there a specific exception that Azure raised when you weren't able to delete these ?

File: app/models/manageiq/providers/azure/cloud_manager/provision/cloning.rb (hunk 4 above)
Comment (liu-samuel) [reply app/models/manageiq/providers/azure/cloud_manager/provision/cloning.rb:186]: Yeah, it was an Azure::Armrest::BadRequestException, so I can change it to 
`rescue Azure::Armrest::BadRequestException => e`

File: app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb (hunk 6 above)
Comment (liu-samuel) [reply app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb:23]: Yeah, this was also a Azure::Armrest::BadRequestException, so I can make it
`rescue Azure::Armrest::BadRequestException`

File: app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb (hunk 5 above)
Comment (agrare) [reply app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb:11]: > why do we need a requeue_phase here as well?

I think we should move the requeue_phase out of delete_instance and into here.  That way we keep all of the signal/requeue state directives in the same method

**Comment by agrare** [comment]
Looks good @liu-samuel, I do think we need to add a spec test for this covering the various scenarios

File: app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb [hunk 7]
```diff
@@ -6,4 +6,23 @@ def customize_destination
 
     signal :post_create_destination
   end
+
+  def clone_failure_cleanup
+    return if phase_context[:clone_options].nil?
```
Comment (Fryguy) [inline app/models/manageiq/providers/azure/cloud_manager/provision/state_machine.rb:11]: Minor but rubocop wants a blank line after guard clauses.

File: spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb [hunk 8]
```diff
@@ -113,4 +116,33 @@
       expect(provision.create_nic(false)).to eql(@nic_object.id)
     end
   end
+
+  context "start_clone_task" do
+    before do
+      @resource_group = FactoryBot.create(:azure_resource_group)
+
+
```
Comment (Fryguy) [inline spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb:124]: blank lines? (accidental commit?)

Comment (agrare) [reply spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb:124]: I asked him to push up whatever he had so I could help him with his failing spec

File: spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb [hunk 9]
```diff
@@ -1,10 +1,18 @@
 require 'azure-armrest'
 
 describe ManageIQ::Providers::Azure::CloudManager::Provision::Cloning do
-  let(:provision) { ManageIQ::Providers::Azure::CloudManager::Provision.new }
+  let(:template) { FactoryBot.create(:azure_image, ext_management_system: ems) }
```
Comment (agrare) [inline spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb:4]:
```suggestion
  let(:template) { FactoryBot.create(:azure_image, :ext_management_system => ems) }
```

File: spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb [hunk 10]
```diff
@@ -1,10 +1,18 @@
...
+  let(:provision) { FactoryBot.create(:miq_provision_azure, :options => {:src_vm_id => template.id}, :miq_request => miq_request) }
+  let(:configuration) {double}
```
Comment (agrare) [inline spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb:6]:
```suggestion
  let(:configuration) {double('azure armrest configuration'}
```

Or something to indicate what the double is for

File: spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb [hunk 11]
```diff
@@ -1,10 +1,18 @@
...
+  let(:vms) {double}
```
Comment (agrare) [inline spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb:7]: Same, something like 'virtual machine service'

File: spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb [hunk 12]
```diff
@@ -113,4 +121,34 @@
       expect(provision.create_nic(false)).to eql(@nic_object.id)
     end
   end
+
+  context "start_clone_task" do
+    before do
+      @resource_group = FactoryBot.create(:azure_resource_group)
+      @dest_name = "test"
+      @clone_options = {clone_options: {name: "test-name", location: "germanycentralwest"}}
```
Comment (agrare) [inline spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb:129]: Same with these two if they aren't referenced elsewhere they can be simple local vars in your before block

File: spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb [hunk 13]
```diff
@@ -113,4 +121,34 @@
...
+      allow(nic_service).to receive(:get).and_return(network_interface)
+      
+      allow(nic_service).to receive(:delete)
+      allow(ip_address_service).to receive(:delete)
+    end
+  
+    it 'deletes the public IP and network interface when a BadRequestException is raised' do
+      provision.start_clone_task
+      expect(ip_address_service).to have_received(:delete)
+      expect(nic_service).to have_received(:delete)
```
Comment (agrare) [inline spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb:151]: Nice

File: spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb [hunk 14]
```diff
@@ -113,4 +121,34 @@
...
+      
+      allow(nic_service).to receive(:delete)
+      allow(ip_address_service).to receive(:delete)
+    end
+  
+    it 'deletes the public IP and network interface when a BadRequestException is raised' do
+      provision.start_clone_task
+      expect(ip_address_service).to have_received(:delete)
+      expect(nic_service).to have_received(:delete)      
+    end
```
Comment (agrare) [inline spec/models/manageiq/providers/azure/cloud_manager/provision/cloning_spec.rb:152]: @liu-samuel can you add a second `it` that checks the phase_context is set properly after BadRequestException is raised ?

**Review by agrare — APPROVED** [review]

**Comment by Fryguy** [comment]
Backported to `radjabov` in commit 87c44f09bcaa077c229f241b9821a102d43422a9.

```text
commit 87c44f09bcaa077c229f241b9821a102d43422a9
Author: Adam Grare <adam@grare.com>
Date:   Tue Sep 10 17:54:35 2024 -0400

    Merge pull request #560 from liu-samuel/clean-up-failed-provision
    
    Clean up network interface and public ip on failure
    
    (cherry picked from commit 2c3a909ec3c7123422c641672c549d1f2d7aa0d2)
```

**Pull request merged**